### DISCUSSION TITLE
Need for threading

### DISCUSSION THREAD

**Discussion by jnoortheen** [xonsh/xonsh#4710] [General]
I have started working on [removing the need for threads](https://github.com/xonsh/xonsh/pull/4634) since there are many problems around using threads for subprocesses. Here is what I found about moving forward to stable subprocess execution

### I. `$XONSH_CAPTURE_ALWAYS = False`
    a. the following modes don't need threading at all since they can block 
        1. `$()` - capture stdout alone with blocking read on the pipe, and pass `sys.stderr`
        2. `$[]` or `![]` - no capture needed. we can pass both stderr/out
    b. object - `!()` 
        - this is implemented as non-blocking. 
        - Here we have to capture both out and err, so we pass the write end to the subprocess. 
        - If we don't read it 
        from xonsh, the subprocess has to wait until the pipe is ready for write (pipes have some max limit like 64kb).
            1. We can avoid the situation by starting a thread as soon as the subprocess starts to start reading.
                But this will also accumulate the data in memory. 
            2. Otherwise we can mention it in doc to tell the user has to call `!().out` if the output is more than
                the pipe's max size.

### II. `XONSH_CAPTURE_ALWAYS=True` and removal of predictable threading
    we need to tee the output for `$[]` or `![]` . 
    we can use the snippet https://stackoverflow.com/questions/41542960 .  
    With this we can remove the need for thread-predictors . we can call any interactive commands 
    and use asyncio.SelectorEventloop (unix) or ProactorEventloop (windows) to make the 
    non-blocking calls simulaneously.  
    But the issue here is that the we can't differentiate stderr and stdout 

### III. Function aliases

I have some ideas about, using [capfd](https://docs.pytest.org/en/6.2.x/capture.html) like redirection for the aliases, and use multiprocessing, when we need to run it in background . But we can discuss about it later, once the subprocesses are sorted out

### TLDR:
    - `!()` is non-blocking, can we make it blocking ?
    - we can remove predictable threading and stream stdin for interactive processes in raw mode. 
        but we can't differentiate stderr/stdout. Can we make this tradeoff?

Let me know what do you think about them.

**Comment by jnoortheen** [comment]
@xonsh/xore you are welcome to comment here . Please use reply in thread when possible

**Reply by gforsyth** [reply]
I don't see any reason why we can't make `!()` blocking.

I think getting rid of threadable predictors would be a _huge_ win and so is worth quite a few tradeoffs.  We'll still have return codes for truthiness checks and so long as we don't lose stream redirection for uncaptured commands (and I don't think we will) that this is worth it.

**Reply by anki-code** [reply]
Okay, after I spent 2 years on fixing bugs and making improvements to xonsh core processes I can comment here.

1. The introduction in first message sounds to me like:
    * We will remove the threading predictor and rewrite huge amount of core logic (with new issues and new edge cases),
    * But the need to run and manage parallel processes/threads will stay (with new issues and new edge cases),
    * And we remove part of key functionality that is working.
    
    This is not good deal to me.

2. After fixing issues I see that our main pain point is 5631. It's the issue about managing stdout/stderr descriptors and proper managing processes in pipe. First of all we need to put all our effort there because solving this is a gold key for other features.

    After I uncovered and fixed issues around Specs I see no reports in the tracker from users about issues. So we can live with threading_predictor for now and put all our energy to improving current xonsh core logic.
    
    Yes, we need more managebale CommandPipepeline. We need better SubprocSpec, etc (hello Slug). But because of we have issues with FDs and we don't understand how callable alias is working with execx all improvements are building on bad platform.
    
    Let's solve 5631 first.

3.  > we can't differentiate stderr/stdout. Can we make this tradeoff?

    I think no because this is the key shell feature.

4. > `!()` is non-blocking, can we make it blocking ?

    Probably yes because I haven't seen good examples of using it. Personally I treat this operator as nice xonsh feature but _as posted by @scopatz in https://github.com/xonsh/xonsh/issues/1758#issuecomment-251032267_

    > I think this mess is partially a result of all of the things that we would like to do that are waaaay beyond the scope of a normal shell.

Finally I'm +1 to any proposals about how we can improve xonsh core. Just one - please make more clear description about pros, cons, trade offs and removed features.

**Comment by shemeshg** [comment]
Is this #4871  related?

**Comment by anki-code** [comment]
@jnoortheen I found the PR that introduced threading and why threading is needed: [Contradiction that is then solved with threads](https://github.com/xonsh/xonsh/pull/1758#issuecomment-251032267). Please take a look.

(I'm going to consolidate the info in https://github.com/xonsh/xonsh/issues/5444)

**Reply by jnoortheen** [reply]
The consolidation is great. What was previously done using threads can now be done with asyncio possibly.

**Reply by anki-code** [reply]
It sounds good. I'm going to take a look into your approach and help where I can after we release awesome 0.17.0.
Thanks for working on this!

**Reply by anki-code** [reply]
I've updated https://github.com/xonsh/xonsh/issues/5444 summary. The "Operators" section is full of tables and links for understanding of how it works now.

**Reply by anki-code** [reply]
In the process of diving into xonsh process management I faced with this example:

```
@aliases.register("mysudo")
def _mysudo():
    input("Password:")
    echo 123

mysudo | grep 1
```
I expect the behavior like `$(sudo -k echo 123 | grep 1)` where you can enter the password and then got captured 123. But it's not working as expected for alias above.

During tracing this case I found that [unthreadable callable alias is not supporting pipes](https://github.com/xonsh/xonsh/issues/5317#issuecomment-2145603549) and I created the group of callable aliases that helps to test and debug the callable alias hebavior - [onsh-developer-toolkit/callias.xsh](https://github.com/anki-code/xonsh-developer-toolkit/blob/main/callias.xsh) - see `_printer`.

As result I have thoughts that first of all we need to describe of what we expect in every case we have. We need to create clear list of test matrix first. I mean something like this [xonsh-cheatsheet/operators](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#operators).

Without clear understanding how it should work I have no sense of true way.

**Comment by anki-code** [comment]
I think what we need before implementation of async is refactoring the logic of how we manage processes in pipe in [`CommandPipeline.iterraw()`](https://github.com/xonsh/xonsh/blob/9e86938c4dc3aacfca989b9e474d78c69394b763/xonsh/procs/pipelines.py#L238). This is the key function that implements waiting for ending and how std streams flow. 

Current logic has issues:

* There are two parts ([1](https://github.com/xonsh/xonsh/blob/9e86938c4dc3aacfca989b9e474d78c69394b763/xonsh/procs/pipelines.py#L258-L282),[2](https://github.com/xonsh/xonsh/blob/9e86938c4dc3aacfca989b9e474d78c69394b763/xonsh/procs/pipelines.py#L297-L345)) and they do the similar things. I think this should work in one loop. See next point.

* We need to wait for the ending processes in pipe much more carefully. Current logic just leave the processes in the background in case of last process ended. Because of it we can see late output, exceptions or endless processes in the background.

These issues lead to:
* https://github.com/xonsh/xonsh/issues/5482
  * See details in [this comment](https://github.com/xonsh/xonsh/issues/5482#issuecomment-2156256492) as well.
* https://github.com/xonsh/xonsh/issues/5317
* maybe it's also related to https://github.com/xonsh/xonsh/issues/4655

As the outcome of this refactoring we need to have clear understanding how we manage pipe and as result where we need to implement async.

**Reply by anki-code** [reply]
@jnoortheen I see in your code that you [refactored iterraw](https://github.com/jnoortheen/xonsh/blob/f8be983a5a944877d58a255051ec1d211deff497/xonsh/procs/pipelines.py#L262).

**Reply by jnoortheen** [reply]
I think it was a refactoring and removal of thread specific ones and wasn't functional.
